backend/packages/Upgrade/src/api/services/MoocletExperimentService.ts

```diff
@@ -1375,6 +1436,120 @@ export class MoocletExperimentService extends ExperimentService {
     return SUPPORTED_MOOCLET_ALGORITHMS.includes(assignmentAlgorithm);
   }
 
+  public async checkForMoocletAssignmentAlgorithmChange(
+    experiment: ExperimentDTO,
+    logger: UpgradeLogger
+  ): Promise<{
+    hasChanged: boolean;
+    wasMooclet: boolean;
+    isNowMooclet: boolean;
+    oldAlgorithm: ASSIGNMENT_ALGORITHM;
+  }> {
+    logger.info({ message: `Check for assignment algorithm change for experiment id => ${experiment.id}` });
+    const oldExperiment = await this.findOne(experiment.id, logger);
+    const hasChanged = oldExperiment.assignmentAlgorithm !== experiment.assignmentAlgorithm;
+    const wasMooclet = this.isMoocletExperiment(oldExperiment.assignmentAlgorithm);
+    const isNowMooclet = this.isMoocletExperiment(experiment.assignmentAlgorithm);
+    return {
+      hasChanged,
+      wasMooclet,
+      isNowMooclet,
+      oldAlgorithm: oldExperiment.assignmentAlgorithm,
+    };
+  }
+
+  public async handlePotentialMoocletAssignmentAlgorithmChange(
+    experiment: ExperimentDTO,
+    currentUser: UserDTO,
+    logger: UpgradeLogger
+  ): Promise<ExperimentDTO> {
+    const { hasChanged, wasMooclet, isNowMooclet, oldAlgorithm } = await this.checkForMoocletAssignmentAlgorithmChange(
+      experiment,
+      logger
+    );
+    let moocletRefToDelete: MoocletExperimentRef | undefined;
+
+    if (hasChanged) {
+      if (experiment.state !== EXPERIMENT_STATE.INACTIVE) {
+        throw new Error(
+          `Experiment state is ${experiment.state}. Can only edit assignment algorithm for experiments yet to begin (INACTIVE state).`
+        );
+      }
+
+      // 1. Cache old mooclet ref BEFORE update, but don't delete stuff from mooclet db yet until after successful update in upgrade db
+      if (wasMooclet) {
+        moocletRefToDelete = await this.getMoocletExperimentRefByUpgradeExperimentId(experiment.id);
+        logger.debug({
+          message: '[Algorithm Change] Fetched mooclet resources for deletion after update',
+          moocletRefId: moocletRefToDelete?.id,
+          experimentId: experiment.id,
+        });
+      }
+
+      // 2. Update experiment with new assignment algorithm, creating new mooclet resources when "isNowMooclet"
+      let message = `[Algorithm Change] Assignment algorithm changed from ${oldAlgorithm} to ${experiment.assignmentAlgorithm}.`;
+      message += wasMooclet && 'Will delete old Mooclet resources after successful update.';
+      message += isNowMooclet && 'Will create new Mooclet resources as part of update.';
+
```

Comment:
Problem: `&&` operator doesn't work as expected for string concatenation. If `wasMooclet` is true, it evaluates to the string, but if false, it adds `false` to the message. Fix:

```
let message = `[Algorithm Change] Assignment algorithm changed from ${oldAlgorithm} to ${experiment.assignmentAlgorithm}.`;
if (wasMooclet) message += ' Will delete old Mooclet resources after successful update.';
if (isNowMooclet) message += ' Will create new Mooclet resources as part of update.';
```